### ISSUE TITLE
Variable inlay type hints not displaying

### ISSUE THREAD

**Issue by camillol** [microsoft/pylance-release#5967]
Type: <b>Bug</b>

Install Pylance and turn on these hints:

```
    "python.analysis.inlayHints.functionReturnTypes": true,
    "python.analysis.inlayHints.variableTypes": true,
```

Try this file:

```
x = 5
def foo():
    y = 10
    return y
```

I get the inlay type hint for the result of foo, but not for x or y.

<img width="930" alt="Screenshot 2024-06-04 at 5 05 35 PM" src="https://github.com/microsoft/pylance-release/assets/196272/7093ec06-c4fd-4373-8ad9-13a54577e49c">

Extension version: 2024.5.1
VS Code version: Code - Insiders 1.90.0-insider (Universal) (89de5a8d4d6205e5b11647eb6a74844ca23d2573, 2024-06-04T01:36:19.247Z)
OS version: Darwin arm64 23.5.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M2 Max (12 x 2400)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|46, 26, 16|
|Memory (System)|32.00GB (0.19GB free)|
|Process Argv|--crash-reporter-id 7b87a5b2-68e7-4d42-ad22-32e3bce2e5e7|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstes627:30244334
vscod805:30301674
vsaa593cf:30376535
py29gd2263:31024238
c4g48928:30535728
vscrpc:30624061
962ge761:30841072
pythongtdpath:30726887
welcomedialog:30812478
pythonidxpt:30768918
pythonnoceb:30776497
asynctok:30898717
dsvsc013:30777762
dsvsc014:30777825
dsvsc015:30821418
pythontestfixt:30866404
pythonregdiag2:30926734
pythonmypyd1:30859725
pythoncet0:30859736
2e7ec940:31000449
pythontbext0:30879054
accentitlementst:30870582
dsvsc016:30879898
dsvsc017:30880771
dsvsc018:30880772
cppperfnew:30980852
pythonait:30973460
jchc7451:31067543
showvideot:31016890
chatpanelt:31014475
9c06g630:31013171
a69g1124:31018687
dvdeprecation:31040973
pythonprt:31036556
dwnewjupyter:31046869
nb_pri_only:31057983
26j00206:31048877

```

</details>

**Comment by debonte** [comment]
This is by design. Starting with 2023.3.22 we stopped showing inlay hints for assignments where the type is readily apparent. This currently includes assignments of literal values, ellipsis, and built-in constants (`False`, `True`, `None`).

Do you feel that there would still be value in showing inlay hints in these scenarios?

Note that if you add a line with a more complex assignment like the following, you will see a variable inlay hint:

```python
x = foo()
```

![image](https://github.com/microsoft/pylance-release/assets/12820898/a5d09052-e41a-47ae-9fb4-e0437d954000)

**Comment by camillol** [comment]
Oh! No, I think that's actually a good idea. I was trying to debug a coworker's issue with type inlays not displaying, and we tried copying a complex expression into a global variable, and it didn't display the inlays; but trying it again now it works, so I think it may have just been pylance being too slow or not having loaded yet. Then I tried a very small test case, which didn't work for a different reason.

I actually wondered if there may not be different settings, e.g. "no inlays / show important inlays / show all inlays", and I looked at the UI for this setting too, but was just a boolean checkbox, so I assumed this was not the case.

I think for maximum discoverability this could be a setting. But unless/until someone asks for "show all" to be an option (I don't actually care myself), it would be enough to add a static text note in the settings, e.g. "Note: inlays are not displayed where the type is obvious, e.g. for literals".

**Comment by debonte** [comment]
@luabud, we're proposing to change the description of the `python.analysis.inlayHints.variableTypes` setting from:

```text
Enable/disable inlay hints for variable types.
```

to something like:

```text
Enable/disable inlay hints for variable types. Hints are never displayed when the type is readily apparent (ex. assignment of a literal).
```

What do you think?

Btw, thoughts on "obvious" vs. "readily apparent" or other options? I'm a little worried that "obvious" could make the user feel bad if it's not obvious to them. 😄

**Comment by rchiodo** [comment]
`Readily apparent` seems off to me. Same reason you didn't pick `obvious`. Why say it all? 

How bout:

```
Hints are not displayed for literal assignments.
```

**Comment by debonte** [comment]
Currently it's not just literals, but also constants (ex. `None`) and ellipsis. And maybe in the future there will be other "obvious" scenarios that we decide to skip.

But yeah, I can see that just saying "literals" is close enough to the truth and simpler to understand.

**Comment by luabud** [comment]
I like the new proposal! We could also include "of literals or constants" if anything

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2024.6.100, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20246100-13-June-2024-PreRelease)
